Add cpp docs for torch::stable (#170912)
pytorch/pytorch: docs/cpp/source/stable.rst

Torch Stable API
================

The PyTorch Stable C++ API provides ABI-stable interfaces for tensor operations.
These functions are designed to maintain binary compatibility across PyTorch versions,
making them suitable for use in ahead-of-time compiled code.

For more information on the stable ABI, see the
`Stable ABI notes <https://docs.pytorch.org/docs/stable/notes/libtorch_stable_abi.html>`_.

Library Registration Macros
---------------------------

These macros provide stable ABI equivalents of the standard PyTorch operator
registration macros (``TORCH_LIBRARY``, ``TORCH_LIBRARY_IMPL``, etc.).
Use these when building custom operators that need to maintain binary
compatibility across PyTorch versions.

``STABLE_TORCH_LIBRARY(ns, m)``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Defines a library of operators in a namespace using the stable ABI.

> This is the stable ABI equivalent of ``TORCH_LIBRARY``. Use this macro to

Reviewer: could we have these link to the original docs, so for this one TORCH_LIBRARY, and below, stable new_empty can redirect to unstable new_empty? That way we could have these docs be minimal + describing differences, where the bulk of the content can still live in the unstable docs.

       m.impl("my_op", TORCH_BOX(my_cuda_kernel));
   }

Minimum compatible version: PyTorch 2.9.

``STABLE_TORCH_LIBRARY_FRAGMENT(ns, m)``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Extends operator definitions in an existing namespace using the stable ABI.

This is the stable ABI equivalent of ``TORCH_LIBRARY_FRAGMENT``. Use this macro
to add additional operator definitions to a namespace that was already
created with ``STABLE_TORCH_LIBRARY``.

**Parameters:**

- ``ns`` - The namespace to extend.
- ``m`` - The name of the StableLibrary variable available in the block.

Minimum compatible version: PyTorch 2.9.

``TORCH_BOX(func)``
^^^^^^^^^^^^^^^^^^^

Wraps a function to conform to the stable boxed kernel calling convention.

This macro takes an unboxed kernel function and generates a boxed wrapper
that can be registered with the stable library API.

**Parameters:**

- ``func`` - The unboxed kernel function to wrap.

**Example:**

.. code-block:: cpp

   Tensor my_kernel(const Tensor& input, int64_t size) {
       return input.reshape({size});
   }

   STABLE_TORCH_LIBRARY_IMPL(my_namespace, CPU, m) {
       m.impl("my_op", TORCH_BOX(my_kernel));
   }

> Minimum compatible version: PyTorch 2.9.

Reviewer: oh yea i added that lol ooooop i forgot

Tensor Class
------------

The ``torch::stable::Tensor`` class provides a stable ABI wrapper around PyTorch
tensors. It offers a user-friendly C++ interface similar to ``torch::Tensor``
while maintaining binary compatibility across PyTorch versions.

.. doxygenclass:: torch::stable::Tensor
   :members:

Device Class
------------

The ``torch::stable::Device`` class provides a stable ABI version of ``c10::Device``.
It represents a compute device (CPU, CUDA, etc.) with an optional device index.

.. doxygenclass:: torch::stable::Device
   :members:

DeviceGuard Class
-----------------

The ``torch::stable::accelerator::DeviceGuard`` class provides a stable ABI version

> of ``c10::DeviceGuard``. It is an RAII class that sets the current device on

Reviewer: same here

construction and restores the previous device on destruction.

.. doxygenclass:: torch::stable::accelerator::DeviceGuard
   :members:

.. doxygenfunction:: torch::stable::accelerator::getCurrentDeviceIndex

Stream Utilities
----------------

> For CUDA stream access, we currently recommend the raw shim API. This will be

Reviewer: ```suggestion
For CUDA stream access, we currently recommend the ABI stable C shim API. This will be
```

improved in a future release with a more ergonomic wrapper.

Getting the Current CUDA Stream
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To obtain the current ``cudaStream_t`` for use in CUDA kernels:

.. code-block:: cpp

   #include <torch/csrc/inductor/aoti_torch/c/shim.h>
   #include <torch/headeronly/util/shim_utils.h>

   // For now, we rely on the raw shim API to get the current CUDA stream.
   // This will be improved in a future release.
   // When using a raw shim API, we need to use TORCH_ERROR_CODE_CHECK to

>    // check the error code and throw an appropriate runtime_error otherwise.

Reviewer: ```suggestion
   // For now, we rely on the ABI stable C shim API to get the current CUDA stream.
   // This will be improved in a future release.
   // When using a C shim API, we need to use TORCH_ERROR_CODE_CHECK to
   // check the error code and throw an appropriate runtime_error otherwise.
```

CUDA Error Checking Macros
--------------------------

These macros provide stable ABI equivalents for CUDA error checking.
They wrap CUDA API calls and kernel launches, providing detailed error
messages using PyTorch's error formatting.

``STD_CUDA_CHECK(EXPR)``
^^^^^^^^^^^^^^^^^^^^^^^^

Checks the result of a CUDA API call and throws an exception on error.
Users of this macro are expected to include ``cuda_runtime.h``.

**Example:**

.. code-block:: cpp

   STD_CUDA_CHECK(cudaMalloc(&ptr, size));
   STD_CUDA_CHECK(cudaMemcpy(dst, src, size, cudaMemcpyDeviceToHost));

Minimum compatible version: PyTorch 2.9.

``STD_CUDA_KERNEL_LAUNCH_CHECK()``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Checks for errors from the most recent CUDA kernel launch. Equivalent to
``STD_CUDA_CHECK(cudaGetLastError())``.

**Example:**

.. code-block:: cpp

   my_kernel<<<blocks, threads, 0, stream>>>(args...);
   STD_CUDA_KERNEL_LAUNCH_CHECK();

Minimum compatible version: PyTorch 2.9.

Header-Only Utilities
---------------------

The ``torch::headeronly`` namespace provides header-only versions of common
PyTorch types and utilities. These can be used without linking against libtorch,

> making them ideal for stable ABI code.

Reviewer: ```suggestion
making them ideal for maintaining binary compatibility across PyTorch versions.
```
